src/pmdas/slurm/pmdaslurm.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2015 Martins Innus
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PMDASLURM 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmdaslurm\f1 \- SLURM statistics performance metrics domain agent (PMDA)
.SH DESCRIPTION
\f3pmdaslurm\f1 is a Performance Metrics Domain Agent (\s-1PMDA\s0) which
exports metric values from the
slurm perlapi
interface to provide information on slurm jobs, nodes and users.
.SH INSTALLATION
Install the slurm PMDA by using the Install script as root:
.sp 1
.RS +4
.ft B
.nf
# cd $PCP_PMDAS_DIR/slurm
# ./Install
.fi
.ft P
.RE
.sp 1
.PP
To uninstall, do the following as root:
.sp 1
.RS +4
.ft B
.nf
# cd $PCP_PMDAS_DIR/slurm
# ./Remove
.fi
.ft P
.RE
.sp 1
\fBpmdaslurm\fR is launched by \fIpmcd\fR(1) and should never be executed
directly. The Install and Remove scripts notify \fIpmcd\fR(1) when the
agent is installed or removed.
.SH FILES
.IP "\fB$PCP_PMDAS_DIR/slurm/Install\fR" 4
installation script for the \fBpmdaslurm\fR agent
.IP "\fB$PCP_PMDAS_DIR/slurm/Remove\fR" 4
undo installation script for the \fBpmdaslurm\fR agent
.IP "\fB$PCP_LOG_DIR/pmcd/slurm.log\fR" 4
default log file for error messages from \fBpmdaslurm\fR
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fR are used to parameterize
the file and directory names used by \fBPCP\fR. On each installation, the
file \fB/etc/pcp.conf\fR contains the local values for these variables.
The \fB$PCP_CONF\fR variable may be used to specify an alternative
configuration file, as described in \fIpcp.conf\fR(5).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pmcd (1)
and
.BR slurm (3).

.\" control lines for scripts/man-spell
.\" +ok+ perlapi {from slurm perlapi interface}
.\" +ok+ pmdaslurm SLURM slurm {from PMDASLURM and pmdaslurm}


CHANGELOG
pcp-6.2.2 (29 May 2024)
    - Work-in-progress, not yet released.

pcp-6.2.1 (10 April 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdauwsgi: (new) instrumentation from uWSGI servers
	pmdalinux: new hugepage metrics from sysfs (fixed size hugepages)
	pmdalinux: new filesys.uuid and filesys.type metrics
	pmdalinux: new TcpExt metrics from /proc/net/netstat
	pmdalinux: new softnet metrics from newer kernel versions
	pmdalinux: new /proc/vmstat khugepaged metrics
	pmdabpf: updates to add eBPF networking metrics for pcp-atop
	pmdaopenmetrics: fix script error diagnostic cascading exception
    - Client tools and utilities:
	pcp2openmetrics: (new) push PCP metrics in OpenMetrics format
	pmcheck: (new) interrogate and control PCP components
	pmcheck: Redis server detection and agent recommendation
	pmcheck: PostgreSQL server detection and agent recommendation
	pmcheck: uwSGI server detection and agent recommendation
	pcp-atopsar: fix TZ mishandling relating to day-crossover
	pcp-dstat: fix an I/O config typo, sum reads and writes
	pcp-htop: sync latest htop-3.3.0 fixes
	pmrep.conf: use bpf instead of bcc pmda for proc net metrics
	pmrep.conf: proc-essential cosmetic improvements
	pmrep.conf: remove vmeff field from sar -B config
	pmrepconf: use pmrep-internal representations for each metric
    - Server-side utilities and log management scripts:
	pmlogger: bizarre timezone fix (github #1936)
	pmlogger: improve handling of a change in metadata semantics
	pmlogdump: add support for multi-archive contexts
	pmproxy: disable Redis protocol proxying by default
	runaspcp: remove hard-coded "pcp" for user and group
	init scripts: drop conditonal use of setpriv(1) or runuser(1)
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix logmeta.c SEGV with V3 archives
	libpcp: fix memleak on repeated __pmFixPMNSHashTab calls
	libpcp: add pmgetopt pmflush() call for warnings
	libpcp: fixes for V3 archives and multi-archive contexts
	libpcp: export __pmLogChangeArchive() interface to tools
	libpcp: fix for PMNS with multi-archive contexts
	libpcp: small derived metrics bug fix (github #1921)
    - Build, infrastructure and packaging updates:
	Linux MX distro support (Debian without systemd)
	riscv: configure pmdabpf_arch value correctly
	tar packages: fix perl modules install in postinstall script
	eBPF updates: add submodule for blazesym, sync to latest
	debian build: add postrm for pcp-zeroconf
	build: removed cppcheck from 'make check', too flakey
	container: Fedora 40 container testing added
	constainer: fixed Fedora base image location (quay.io)
	build: add support for riscv64 to pcp rpm spec files
    - Security Enhanced Linux:
	selinux policy: small tweak for pmie signal handling
    - Documentation:
	man pages: document pmlogger_check and pmlogger_daily SaveLogs
	man pages: man-spell and corrections
	docs: updated donations page with new fiscal sponsor details

pcp-6.2.0 (12 February 2024)
    - Client tools and utilities:
	pmlogger: default to creating version 3 PCP archives
	pmie, pmlogger, pmcd, pmproxy: init script hardening to
	    run less as root:root, more as pcp:pcp and more safely
	pcp-reboot-init: new systemd service for $PCP_RC_DIR setup
	pcp2arrow: new, export metrics to parquet format for pandas
	pcp-atop: deterministic process sort when handling NULLs
	pcp-atop: update to v2.10.0 upstream code
	pcp-htop: update to latest upstream fixes
	pcp-geolocate: function with older versions of python
	pcp-meminfo: resolve possible 'broken pipe' diagnostics
	pcp-ss: fix handling of listening state sockets to match ss
	pcp-ss: fix the UDP UNCONN vs LISTEN state handling
    - PMDA additions, enhancements and bug fixes:
	pmdaproc: improvements for process accounting metrics
	pmdalinux: add pgpromote* and pgdemote* memory metrics
	pmdalinux: add metric sysfs.module.zswap.max_pool_percent
	pmdalinux: add new SNMP metric network.ip.outtransmits
	pmdasummary: send state change 0 (no change) back to pmcd
	pmdapostgresql: support the psycopg3 python module as well
	pmdaopenmetrics: kepler support - live metrics and logging
    - Server-side utilities and log management scripts:
	pmlogrewrite, pmlogredact: fix for missing data volumes
	pmcd: cosmetic change to avoid pipe2 name clash
	runaspcp: new helper utility for unprivileged rc scripts
	pmnsmerge: be more defensive in qsort comparison routine
	pmlogsize: fix qsort comparison function handling of equality
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: pmOpenLog refactor for handling previous log files
	libpcp: new __pmCleanMapDir routine to aid log culling
	libpcp & pmcd: READY-NOT-READY hardening
	libpcp: avoid possible thread issues (coverity)
	libpcp: refactor pmDupContext for thread safety
	libpcp_import: add guard in pmiPutMark to prevent SIGSEGV
    - Build, infrastructure and packaging updates:
	build: changes to re-enable builds for FreeBSD 12, 13, 14
	build: add diffutils rpm dep as pmdaproc script uses diff
	build: add main.yml for building and publishing container
	build: resolve compiler warnings in various tools/libraries
	packaging: use tmpfiles.d method to create PCP_RUN_DIR
	build: packaging changes for openSUSE 15.5 and 15.6
	debian: do not ship empty directory /usr/lib/pkgconfig
	build: el10 systemd-presets for pmcd/pmie/pmlogger enabling
	build: drop rpm dep on bpftool as local version used only
	build: change perl YAML::XS::LibYAML module to YAML::XS
    - Security Enhanced Linux:
	selinux: policy improvements for CentOS/RHEL 7 and 8
    - Documentation:
	docs: Fedora 39 found some tbl(1)/troff(1) issues, fix
	docs: add an explicit code-of-conduct for PCP
	docs: pmns(5) note that metric names must only contain ASCII
	docs: update the BPF README to describe bpftool reality

pcp-6.1.1 (17 November 2023)
    - Client tools and utilities:
	pmie: add -o/--format to control output format with archives

src/pmdas/slurm/pmdaslurm.pl
#
# Copyright (c) 2015 Martins Innus.
# All rights reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#

use strict;
use warnings;

use PCP::PMDA;
use Net::Domain qw( hostname );

use Slurm;
use Slurm::Hostlist;
use Slurm qw(:constant);

use threads;
use threads::shared;

use vars qw( $pmda $slurm );


# indoms
my ( $slurm_indom, $node_job_indom, $job_indom, $user_indom, $node_indom ) = ( 0, 1, 2, 3, 4);

my $GEN_CLUSTER = 0;
my $NODE_JOB_CLUSTER = 1;
my $GLOBAL_JOB_CLUSTER = 2;
my $NODE_CLUSTER = 3;
my $USER_CLUSTER = 4;
my $PARTITION_CLUSTER = 5;

# Slurm parameters we actually care about
# Order matches the metric item number to make the fetch logic easier
# Names match the struct entries in slurm.h:job_info if existing, otherwise prepend with "DERIVED"

my @slurm_job_stats = ( "FILLER_FOR_NUM_JOBS", "job_id", "name", "job_state", "user_id", "batch_host", "submit_time", "start_time", "end_time", "features", "gres", "nodes", "num_nodes", "num_cpus", "ntasks_per_node", "work_dir", "DERIVED_job_alloc_cores");

#
# Shared Variables between slurm gather thread and main pmda thread
#
# Will lock on the variable we want to use.

our %nodejobs :shared = ();
our $numnodes :shared = 0;

#
# End Shared Variables
#

#
# Subs for the slurm polling thread
#

# Who am i. To check for jobs in this host
our $host = hostname;

# Don't need to share these.  Only used in the slurm thread
my $jobs_update_time = 0;
sub slurm_init {
   $slurm = Slurm::new();
}

# Update General Stuff
sub slurm_update_cluster_gen {

    # This could take a while
    # Returns a hash ref that mirrors : node_info_msg_t from slurm.h
    my $nodemsg = $slurm->load_node();

    unless($nodemsg) {
        # This can fail if the slurm controller has not come up yet
        # No good way to figure out if this is just a delay or fatal error
        # So just log it and try again later
        warn "Failed to load node info: " . $slurm->strerror();
        return;
    }

    my @nodes = @{ $nodemsg->{node_array} };

    {
        lock($numnodes);
        $numnodes = @nodes;
    }

}

# Update job information for this node
sub slurm_update_cluster_jobs {

    # Use the time so we only update on changes
    #
    # This could take a while
    # Returns a hash ref where the main element we care about is the job_array array
    my $jobflags = Slurm::SHOW_DETAIL;
    my $jobmsg = $slurm->load_jobs($jobs_update_time, $jobflags);

    unless($jobmsg) {
        # This can fail if the slurm controller has not come up yet
        # Or if there is no state change from the previous try.
        # If there is no state change, the main thread will just use existing data.
        # No good way to figure this out
        # So just try again later
        return;
    }

    # Grab the update time to use for the next query
    $jobs_update_time = $jobmsg->{last_update};

    %nodejobs = ();

    {
        lock(%nodejobs);

        for my $job ( @{ $jobmsg->{job_array} } ){

            my $jid = $job->{job_id};
            my $nodelist = $job->{nodes};

            if( $nodelist){

                my $hostlist = Slurm::Hostlist::create( $nodelist );
                my $num_hosts = $hostlist->count();
            
                if ( $num_hosts > 0){
                    my $mypos = $hostlist->find($host);
                    if( $mypos >= 0){
                        # We only want jobs that are on this host
                        if ( Slurm::IS_JOB_RUNNING($job) ){
                            # And Running
                       
                            # Ugh, shared perl hashes
                            $nodejobs{$jid} = &share( {} );
                            while ( my ($key, $value) = each(%$job)){
                                # Only grab the things we really want, to keep memory down
                                if ( grep{$_ eq $key} @slurm_job_stats ){
                                    $nodejobs{$jid}{$key} = $value;
                                }
                            }
                            #
                            # Grab the cores allocated per node
                            # Not exposed in the API so we need to parse the string representation of the job
                            my $slurmstr = $slurm->sprint_job_info($job,1);
                            my $nodecores = "";
                            # Should always exist, but just in case
                            # Could be multiple matches
                            # SLURM api groups hosts together that have the same cores allocated
                            while ( $slurmstr =~ /Nodes=(\S*) CPU_IDs=(\S*)/g ){
                                my $foundnodes = $1;
                                my $foundcores = $2;
                                my $corehosts = Slurm::Hostlist::create( $foundnodes );
                                while( my $corehost = $corehosts->shift()){
                                    $nodecores .= "$corehost:$foundcores ";
                                }
                            }
                            $nodecores =~ s/\s+$//;
                            $nodejobs{$jid}{"DERIVED_job_alloc_cores"} = "$nodecores";
                        }
                    }
                }
            }
        }
    } # unlock %nodejobs
}

# slurm fetch worker thread
sub poll_slurm {
    while (1){
        slurm_update_cluster_gen();
        slurm_update_cluster_jobs();
        sleep 9;
    }
}

#
# Subs for the main thread
#

#
# fetch is called once by pcp for each refresh and then the fetch callback is
# called to query each statistic individually
#
sub slurm_fetch {

        my @nodejobs_array;

        {
            lock(%nodejobs);

            # Create a [key1, "key1", key2, "key2", .....] array so we can control the inst ids
            while ( my ($key, $val) = each %nodejobs){
                push( @nodejobs_array, $key );
                push( @nodejobs_array, "$key" );
            }
        } # unlock %nodejobs

        $pmda->replace_indom($node_job_indom, \@nodejobs_array);
}

sub slurm_fetch_node_job_callback {
    my ($item, $inst) = @_;

    if ( $item == 0 ){
        my $numjobs;
        {
            lock(%nodejobs);
            $numjobs = keys %nodejobs;
        }
        return ($numjobs, 1);
    }

    my $lookup;
    {
        lock(%nodejobs);
        # Will unlock on scope loss, return or otherwise
        if( !exists $nodejobs{$inst} ){
            return (PM_ERR_INST, 0);
        }
        else {
            $lookup = $nodejobs{$inst};
        }
    }

    if ( $item == 3 ){
        # slurm.node.job.state
        # List all jobs as running since those are the only jobs we grabbed
        return ("running", 1);
    }
    elsif ( $item < scalar @slurm_job_stats ){
        my $rv = $lookup->{ $slurm_job_stats[$item] };
        $rv = "" if !defined $rv;
        return ( $rv, 1);
    }
    else{
        return (PM_ERR_PMID, 0);
    }
}

sub slurm_fetch_callback {
    my ($cluster, $item, $inst) = @_; 

    my $pmid_name = pmda_pmid_name($cluster, $item)
        or return (PM_ERR_PMID, 0);
        #or die "Unknown metric name: cluster $cluster item $item\n";

    if( $cluster == 0 && $item == 0 ){
        return ($numnodes, 1);
    }
    elsif( $cluster == $NODE_JOB_CLUSTER){
        return slurm_fetch_node_job_callback( $item, $inst );
    }
    else{
        return (PM_ERR_PMID, 0);
    }
}

# the PCP::PMDA->new line is parsed by the check_domain rule of the PMDA build
# process, so there are special requirements:  no comments, the domain has to
# be a bare number.
#
our $pmda = PCP::PMDA->new('slurm', 23);

# metrics go here, with full descriptions

# general - cluster 0
$pmda->add_metric(pmda_pmid($GEN_CLUSTER,0), PM_TYPE_U32, PM_INDOM_NULL,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.numnodes',
    'Number of slurm nodes',
    '');

# To add - cluster name, controller name, slurm version

# node job cluster 
$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,0), PM_TYPE_U32, PM_INDOM_NULL,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.numjobs',
    'Number of jobs running on this node',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,1), PM_TYPE_U32, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.id',
    'Slurm jobid',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,2), PM_TYPE_STRING, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.name',
    'Job name',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,3), PM_TYPE_STRING, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.state',
    'Job state',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,4), PM_TYPE_U32, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.user_id',
    'UID owning this job',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,5), PM_TYPE_STRING, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.batch_host',
    'First host in the allocation, the one running the batch script',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,6), PM_TYPE_STRING, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.submit_time',
    'Time the job was submitted',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,7), PM_TYPE_STRING, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.start_time',
    'Time the job started',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,8), PM_TYPE_STRING, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.end_time',
    'Time the job ended',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,9), PM_TYPE_STRING, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.features',
    'Features requested by the job',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,10), PM_TYPE_STRING, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.gres',
    'Generic resources (gres) requested by the job',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,11), PM_TYPE_STRING, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.nodes',
    'Nodes allocated to the job',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,12), PM_TYPE_U32, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.num_nodes',
    'Number of nodes requested by the job',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,13), PM_TYPE_U32, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.num_cpus',
    'Total cpus requested by the job',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,14), PM_TYPE_U32, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.ntasks_per_node',
    'Number of tasks per node requested by the job',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,15), PM_TYPE_STRING, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.work_dir',
    'Working directory of the job',
    '');

$pmda->add_metric(pmda_pmid($NODE_JOB_CLUSTER,16), PM_TYPE_STRING, $node_job_indom,
    PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
    'slurm.node.job.job_alloc_cores',
    'Cores allocated to the job',
    '');

# To add - batch script, alloc_node, account, partition ?

# global job cluster 
#
# ( priorities for queued jobs, running jobs keep last priortiy)
# Need to add priority func to api/job_info.c take from sprio.c
# then perl wrap.  need to use list fcns

# user cluster

# node cluster

# per cluster cluster - may mirror metrics in general above ???

&slurm_init;

# Thread that queries the slurm state.
#
# Slurm calls can block for a while so need a seperate thread
# so pmcd doesn't kill us if fetch's take too long.

my $slurm_thread = threads->create(\&poll_slurm);
$slurm_thread->detach();

$node_job_indom = $pmda->add_indom($node_job_indom, [], '', '');

$pmda->set_fetch(\&slurm_fetch);
$pmda->set_fetch_callback(\&slurm_fetch_callback);

$pmda->run;
